autopilot/contractor.go

```diff
@@ -140,28 +145,39 @@ func (c *contractor) performContractMaintenance(cfg Config) error {
 		return err
 	}
 
-	// fetch all active contracts
-	active, err := c.ap.bus.Contracts()
+	// fetch all active contracts and their latest revisions.
+	contractData, err := c.ap.bus.Contracts()
+	if err != nil {
+		return err
+	}
+	revisions, err := c.ap.worker.Revisions(c.ap.masterKey, contractData)
 	if err != nil {
 		return err
 	}
+	var contracts []contract
+	for i, c := range contractData {
+		if c.ID != revisions[i].ID() {
+			panic("shouldn't happen")
+		}
+		contracts = append(contracts, contract{Contract: c, ContractRevision: revisions[i]})
+	}
 
 	// run checks
-	toDelete, toIgnore, toRenew, err := c.runContractChecks(cfg, active)
+	toDelete, toIgnore, toRenew, err := c.runContractChecks(cfg, contracts)
 	if err != nil {
 		return fmt.Errorf("failed to run contract checks, err: %v", err)
 	}
 
 	// delete contracts
 	if len(toDelete) > 0 {
-		err = c.ap.bus.DeleteContracts(contractIds(toDelete))
+		err = c.ap.bus.DeleteContracts(toDelete)
 		if err != nil {
 			return fmt.Errorf("failed to delete contracts, err: %v", err)
 		}
 	}
 
 	// find out how much we spent in the current period
-	spent, err := c.currentPeriodSpending()
+	spent, err := c.currentPeriodSpending(contracts)
```

Comment:
`currentPeriodSpending` and `contractSpending` is something I eventually want to move over to the bus. SQL should be able to fetch contract hierarchies using recursive queries